tools/buildsteps/win32/buildffmpeg.sh

```diff
@@ -0,0 +1,183 @@
+FFMPEG_DEFAULT_OPTS=""
+
+BASE_URL=$(grep "BASE_URL=" $FFMPEG_VERSION_FILE | sed 's/BASE_URL=//g')
+VERSION=$(grep "VERSION=" $FFMPEG_VERSION_FILE | sed 's/VERSION=//g')
+LIBNAME=ffmpeg
+ARCHIVE=$(echo "${VERSION}" | sed 's/\//-/g').tar.gz
+FFMPEGDESTDIR=/xbmc/lib/win32/$LIBNAME
+
+[[ -f buildhelpers.sh ]] &&
+    source buildhelpers.sh
+
+do_getFFmpegConfig() {
+  if [[ -f "$FFMPEG_CONFIG_FILE" ]]; then
+    FFMPEG_OPTS_SHARED="$FFMPEG_BASE_OPTS $(cat "$FFMPEG_CONFIG_FILE" | sed -e 's:\\::g' -e 's/#.*//')"
+  else
+    FFMPEG_OPTS_SHARED="$FFMPEG_BASE_OPTS $FFMPEG_DEFAULT_OPTS"
+  fi
+
+  if [[ $bits = "64bit" ]]; then
+    arch=x86_64
+  else
+    arch=x86
+  fi
+  export arch
+
+  # add options for static modplug
+  if do_checkForOptions "--enable-libmodplug"; then
+    do_addOption "--extra-cflags=-DMODPLUG_STATIC"
+  fi
+
+  # handle gplv3 libs
+  if do_checkForOptions "--enable-libopencore-amrwb --enable-libopencore-amrnb \
+    --enable-libvo-aacenc --enable-libvo-amrwbenc"; then
+    do_addOption "--enable-version3"
+  fi
+
+  do_removeOption "--enable-nonfree"
+  do_removeOption "--enable-libfdk-aac"
+  do_removeOption "--enable-nvenc"
+  do_removeOption "--enable-libfaac"
+
+  # remove libs that don't work with shared
+  do_removeOption "--enable-decklink"
+  do_removeOption "--enable-libutvideo"
+  do_removeOption "--enable-libgme"
+}
+
+do_checkForOptions() {
+  local isPresent=1
+  for option in "$@"; do
+    for option2 in $option; do
+      if echo "$FFMPEG_OPTS_SHARED" | grep -q -E -e "$option2"; then
+        isPresent=0
+      fi
+    done
+  done
+  return $isPresent
+}
+
+do_addOption() {
+  local option=${1%% *}
+  local shared=$2
+  if ! do_checkForOptions "$option"; then
+    FFMPEG_OPTS_SHARED="$FFMPEG_OPTS_SHARED $option"
+  fi
+}
+
+do_removeOption() {
+  local option=${1%% *}
+  FFMPEG_OPTS_SHARED=$(echo "$FFMPEG_OPTS_SHARED" | sed "s/ *$option//g")
+}
+
+if [ $NUMBER_OF_PROCESSORS > 1 ]; then
+  if [ $NUMBER_OF_PROCESSORS > 4 ]; then
+    cpuCount=6
+  else
+    cpuCount=`expr $NUMBER_OF_PROCESSORS + $NUMBER_OF_PROCESSORS / 2`
+  fi
+fi
+
+do_getFFmpegConfig
+
+if [[ "$tools" = "msvc" ]]; then
+  # this experimental feature for debuging purpose
+  do_removeOption "--enable-libdcadec"
+  do_removeOption "--enable-gnutls"
+  do_removeOption "--disable-debug"
+  do_addOption "--disable-gnutls"
+  do_addOption "--enable-debug"
+  do_addOption "--toolchain=msvc"
+
+  PATH="/c/Program Files (x86)/Microsoft Visual Studio 12.0/VC/BIN/":$PATH
+  CFLAGS=""
+  CXXFLAGS="" 
+  LDFLAGS=""
+  export PATH CFLAGS CXXFLAGS LDFLAGS
+
+  extra_cflags="-MDd -I$LOCALDESTDIR/include -I/mingw32/include"
+  extra_ldflags="-L$LOCALDESTDIR/lib -L/mingw32/lib /NODEFAULTLIB:libcmt"
+  TARGET_OS=win32
+fi
+
+cd $LOCALBUILDDIR
+
+if do_checkForOptions "--enable-gnutls"; then
+  [[ -z "$gnutls_ver" ]] && gnutls_ver=$(curl -sl "ftp://ftp.gnutls.org/gcrypt/gnutls/v3.4/")
```

Comment:
please don't hardcode upstream URLs here. Should point to our mirrors.
same on L122